connman/src/service.c

```diff
@@ -5746,12 +5746,27 @@ static gboolean connect_timeout(gpointer user_data)
 
 	DBG("service %p", service);
 
-	service->timeout = 0;
-
 	if (service->network)
 		__connman_network_disconnect(service->network);
-	else if (service->provider)
+	else if (service->provider) {
+		/*
+		 * Do not remove timeout when the VPN is waiting for user input
+		 * in association state. By default the VPN agent timeout is
+		 * 300s whereas default connection timeout is 120s.
+		 */
+		const char *statestr = connman_provider_get_string(
+					service->provider, "State");
+		if (!g_strcmp0(statestr, "association")) {
+			DBG("VPN provider %p is waiting for user input, "
+						"delay connect timeout",
+						service->provider);
+			return G_SOURCE_CONTINUE;
+		}
+
 		connman_provider_disconnect(service->provider);
+	}
+
+	service->timeout = 0;
```

Comment:
If I'm understanding this code correctly, the timer is started when the association state is entered. While it's in this state, the timer fires every 120 seconds but is ignored.

But in this case, it means there is less time for the connection to complete after the credentials have been entered (suppose it takes _x_ seconds to enter the credentials, then there would be (120 - (_x_ mod 120)) seconds remaining for the connection to complete, or thereabouts).

During testing if I got the timing just right (pressed the "Connect" button on the credentials dialog after 120 seconds), I found the connection failed.

I might be misunderstanding what's going on here, but assuming this is what's happening, would it make sense to reset the timer after the association state has been left, so that there's the full 120 seconds left to complete the connection?

Reply:
Yeah, I guess the timer should be started only when in connect state. I'll see what is to be done with this.

Reply:
0ca49c52a1e89045b5d59e4a1e645804e17a996f and 852d02ff1a95f647ff98fd9b041f0dc76f4971a5 do address the issue. However, I'm now guessing that the VPN entering `CONFIGURATION/CONNECT` state should always restart the timer if there is the case that a plugin needs to use agent multiple times and, thus, goes connect -> association -> connect state. Other way would be just to prevent such transitions... and that would require a bit of a though to the vpn-agent.c change 607c87cbda99655da4525dc7bfdcc897649fbe0f

Reply:
I guess this was ok according to https://github.com/sailfishos/connman/pull/34#pullrequestreview-1222649675 - resolving.